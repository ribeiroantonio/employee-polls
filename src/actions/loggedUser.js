export const LOG_USER_IN = "LOG_USER_IN";

const logUserIn = (id) => {
  return {
    type: LOG_USER_IN,
    id,
  };
};

export default logUserIn;
