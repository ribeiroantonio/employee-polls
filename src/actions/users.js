export const RETRIEVE_USERS = "RETRIEVE_USERS";
export const ADD_ANSWER_TO_USER = "ADD_ANSWER_TO_USER";
export const ADD_QUESTION_TO_USER = "ADD_QUESTION_TO_USER";

export const retrieveUsers = (users) => {
  return {
    type: RETRIEVE_USERS,
    users,
  };
};

export const addAnswerToUser = ({ loggedUser, questionId, answer }) => {
  return {
    type: ADD_ANSWER_TO_USER,
    loggedUser,
    questionId,
    answer,
  };
};

export const addQuestionToUser = ({ id, author }) => {
  return {
    type: ADD_QUESTION_TO_USER,
    id,
    author,
  };
};
