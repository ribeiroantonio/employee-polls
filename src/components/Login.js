import React, {useState} from "react";
import { connect } from "react-redux";
import { useLocation, useNavigate } from "react-router-dom";
import logUserIn from "../actions/loggedUser";

const Login = ({ users, dispatch }) => {
  const [name, setName] = useState("");
  const [password, setPassword] = useState("");

  const navigate = useNavigate();
  const location = useLocation();

  const validateCredentials = () => {
    return users.find((user) => user.id === name && user.password === password);
  }

  const handleLogin = (e) => {
    e.preventDefault();

    if (validateCredentials()) {
      dispatch(logUserIn(name));
      const path = location.pathname;
      if (path.includes("login")) {
        navigate("/");
      } else {
        navigate(path);
      }
    } else {
      alert("Invalid Credentials");
    }
  };

  return (
    <div className="container-sm mt-5">
      <form>
        <h2>Log in to Employee Polls</h2>
        <div className="mb-3">
          <label htmlFor="userNameInput" className="form-label">Username</label>
          <input type="text" className="form-control" id="userNameInput" value={name} onChange={(e) => setName(e.target.value)}/>
        </div>
        <div className="mb-3">
          <label htmlFor="passwordInput" className="form-label">Password</label>
          <input type="password" className="form-control" id="passwordInput" value={password} onChange={(e) => setPassword(e.target.value)}/>
        </div>
        <button type="submit" className="btn btn-primary" onClick={handleLogin}>Log In</button>
      </form>
    </div>
  );
};

const mapStateToProps = ({ users }) => {
  return {
    users: Object.keys(users).map((id) => users[id]),
  };
};

export default connect(mapStateToProps)(Login);
