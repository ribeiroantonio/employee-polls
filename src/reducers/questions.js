import {
  RETRIEVE_QUESTIONS,
  SAVE_QUESTION,
  SAVE_ANSWER,
} from "../actions/questions";

export default function questions(state = {}, action) {
  switch (action.type) {
    case RETRIEVE_QUESTIONS:
      return {
        ...state,
        ...action.questions,
      };
    case SAVE_QUESTION:
      return {
        ...state,
        ...questions,
        [action.question.id]: action.question,
      };
    case SAVE_ANSWER:
      const { loggedUser, questionId, answer } = action;
      return {
        ...state,
        [questionId]: {
          ...state[questionId],
          [answer]: {
            ...state[questionId][answer],
            votes: state[questionId][answer].votes.concat([loggedUser]),
          },
        },
      };
    default:
      return state;
  }
};
