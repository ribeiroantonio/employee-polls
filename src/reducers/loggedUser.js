import { LOG_USER_IN } from "../actions/loggedUser";

export default function loggedUser(state = null, action) {
  switch (action.type) {
    case LOG_USER_IN:
      return action.id;
    default:
      return state;
  }
};
