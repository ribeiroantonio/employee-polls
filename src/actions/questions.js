import { saveQuestion } from "../utils/api";
import { addQuestionToUser } from "./users";

export const RETRIEVE_QUESTIONS = "RETRIEVE_QUESTIONS";
export const SAVE_QUESTION = "SAVE_QUESTION";
export const SAVE_ANSWER = "SAVE_ANSWER";

export const retrieveQuestions = (questions) => {
  return {
    type: RETRIEVE_QUESTIONS,
    questions,
  };
};

export const saveUserQuestion = (question) => {
  return {
    type: SAVE_QUESTION,
    question,
  };
};

export const handleAddQuestion = (question) => {
  return (dispatch) => {
    return saveQuestion(question).then((question) => {
      dispatch(saveUserQuestion(question));
      dispatch(addQuestionToUser(question));
    });
  };
};

export const saveQuestionAnswer = ({ loggedUser, questionId, answer }) => {
  return {
    type: SAVE_ANSWER,
    loggedUser,
    questionId,
    answer,
  };
};
