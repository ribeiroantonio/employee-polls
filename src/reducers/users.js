import { RETRIEVE_USERS } from "../actions/users";
import { ADD_ANSWER_TO_USER } from "../actions/users";
import { ADD_QUESTION_TO_USER } from "../actions/users";

export default function users (state = {}, action) {
  switch (action.type) {
    case RETRIEVE_USERS:
      return {
        ...state,
        ...action.users,
      };
    case ADD_QUESTION_TO_USER:
      const { id, author } = action;
      return {
        ...state,
        [author]: {
          ...state[author],
          questions: state[author].questions.concat([id]),
        },
      };
    case ADD_ANSWER_TO_USER:
      const { loggedUser, questionId, answer } = action;
      return {
        ...state,
        [loggedUser]: {
          ...state[loggedUser],
          answers: {
            ...state[loggedUser].answers,
            [questionId]: answer,
          },
        },
      };
    default:
      return state;
  }
};
