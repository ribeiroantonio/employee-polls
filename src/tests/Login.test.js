import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { configureStore } from '@reduxjs/toolkit';
import loggedUser from  '../reducers/loggedUser';
import questions from  '../reducers/questions';
import users from  '../reducers/users';
import Login from "../components/Login";

const store = configureStore({
  reducer: {
    loggedUser,
    questions,
    users,
  },
});

describe("when rendering login page", () => {
  it("matches snapshot", () => {
    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <Login />
        </Provider>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
