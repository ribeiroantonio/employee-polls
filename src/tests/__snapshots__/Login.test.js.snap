// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when rendering login page matches snapshot 1`] = `
<div>
  <div
    class="container-sm mt-5"
  >
    <form>
      <h2>
        Log in to Employee Polls
      </h2>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="userNameInput"
        >
          Username
        </label>
        <input
          class="form-control"
          id="userNameInput"
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="passwordInput"
        >
          Password
        </label>
        <input
          class="form-control"
          id="passwordInput"
          type="password"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Log In
      </button>
    </form>
  </div>
</div>
`;
